component/prometheus/kubernetes/podmonitors/crdManager.go

```diff
@@ -0,0 +1,189 @@
+		if err != nil {
+			level.Error(c.logger).Log("name", pm.Name, "err", err, "msg", "error generating scrapeconfig from podmonitor")
+			continue
+		}
+		c.mut.Lock()
+		c.discoveryConfigs[pmc.JobName] = pmc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[pmc.JobName] = pmc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+func (c *crdManager) onAddPodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	level.Info(c.logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.addPodMonitor(pm)
+}
+func (c *crdManager) onUpdatePodMonitor(oldObj, newObj interface{}) {
+	pm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.addPodMonitor(newObj.(*v1.PodMonitor))
+}
+func (c *crdManager) onDeletePodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.apply()
+}
+
+func (c *crdManager) run(ctx context.Context) error {
+	config, err := c.config.restConfig()
+	if err != nil {
+		return errors.Wrap(err, "creating rest config")
+	}
+	promClientset, err := versioned.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "creating prometheus clientset")
+	}
+
+	c.cg = configGenerator{
+		config: c.config,
+	}
+
+	for _, namespace := range c.config.Namespaces {
+		factory := informers.NewSharedInformerFactoryWithOptions(promClientset,
+			5*time.Minute,
+			informers.WithNamespace(namespace),
+			informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+				if c.config.FieldSelector != "" {
+					opts.FieldSelector = c.config.FieldSelector
+				}
+				if c.config.LabelSelector != "" {
+					opts.LabelSelector = c.config.LabelSelector
+				}
+			}))
+
+		pminf := factory.Monitoring().V1().PodMonitors().Informer()
+		pminf.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAddPodMonitor,
+			UpdateFunc: c.onUpdatePodMonitor,
+			DeleteFunc: c.onDeletePodMonitor,
+		})
+		err = pminf.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+			level.Error(c.logger).Log("msg", "pod monitor watcher error", "err", err)
+		})
+		if err != nil {
+			return errors.Wrap(err, "setting watch handler")
+		}
+
+		factory.Start(ctx.Done())
+	}
+
+	level.Info(c.logger).Log("msg", "informers  started")
+
+	c.discovery = discovery.NewManager(ctx, c.logger, discovery.Name(c.opts.ID))
+	go func() {
+		err := c.discovery.Run()
+		if err != nil {
+			level.Error(c.logger).Log("msg", "discovery manager stopped", "err", err)
+		}
+	}()
+
+	flowAppendable := prometheus.NewFanout(c.config.ForwardTo, c.opts.ID, c.opts.Registerer)
+	opts := &scrape.Options{}
+	c.scraper = scrape.NewManager(opts, c.logger, flowAppendable)
+	defer c.scraper.Stop()
+	targetSetsChan := make(chan map[string][]*targetgroup.Group)
+	go func() {
+		err := c.scraper.Run(targetSetsChan)
+		level.Info(c.logger).Log("msg", "scrape manager stopped")
+		if err != nil {
+			level.Error(c.logger).Log("msg", "scrape manager failed", "err", err)
+		}
+	}()
```

Comment:
There's some risk here that changing the component is going to inject a lot of staleness markers. Ideally there's one scraper shared throughout the lifetime of the component.

Otherwise, from my current understanding the Prometheus scraper, when you shut the scraper down, staleness markers get propagated through the system for every target, even if that target isn't going away, which will cause gaps in Grafana.